storage/src/main/scala/uk/ac/wellcome/storage/transfer/s3/S3Transfer.scala

```
@@ -96,10 +88,14 @@ class S3Transfer(implicit s3Client: AmazonS3)
       Left(TransferOverwriteFailure(src, dst))
     }
 
-  private def getStream(location: S3ObjectLocation): Try[S3ObjectInputStream] =
-    Try {
+  private def getStream(location: S3ObjectLocation) = {
+    val errorOrStream = Try {
       s3Client.getObject(location.bucket, location.key)
-    }.map { _.getObjectContent }
+    }.toEither.left.map(error => S3Errors.readErrors(error)).map {
+      _.getObjectContent
+    }
+    errorOrStream.retry(maxAttempts = 3)
```

Maybe we could replace this with an S3StreamReadable or S3StreamStore? That would give us the retrying functionality “for free” and is already the correct type.

The reason this code wasn't using it before is because S3StreamStore returns an Either[_] rather than a Try[_], and I didn't think it worth doing an Either -> Try conversion. Doing a Try -> Either conversion is a bit icky as well.

(If you inject an S3StreamStore, I think you can also ditch the implicit S3 client from the constructor, which IMO is clearer. Right now one argument can be derived from the other, which seems weird.)

---

Why is a Try -> Either (or the other way around) a bit icky?

---

Your comment makes sense though. Will do

---

> Why is a Try -> Either (or the other way around) a bit icky?

I always find it a bit odd if you create an Either/Try then immediately cast to something else in the same line, because why not go straight to the thing you wanted? But it might just be a style thing.

---

In this case you can't go straight to `Either` because `s3Client.getObject(location.bucket, location.key)` thows an Exception and `Try` is the only thing that allows you to deal with it. I find that when we deal with Java libraries, where the pattern is usually to thow and let the exception bubble up, then we have to go via `Try`. `Either` and `Try` have a lot in common but aren't quite the same and sometimes only one of the 2 will do, so I don't really see anything wrong with going back and forth and using the one that is more convenient in that moment